### DISCUSSION TITLE
etags in Micronaut

### DISCUSSION THREAD

**Discussion by prabhuveerappan** [micronaut-projects/micronaut-core#10980] [Q&A] [answered]
I am trying to work on etags in Micronaut and started with small Item code using the below codes,

mn create-app --build=gradle --jdk=17 --lang=java --test=junit com.example.demo

Added class ItemController.java like below,

package com.example;

import io.micronaut.http.HttpHeaders;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.PathVariable;
import io.micronaut.http.annotation.Produces;
import jakarta.inject.Inject;

@Controller("/items")
public class ItemController {

    @Inject
    private ItemService itemService;

    @Get("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public HttpResponse<Item> getItem(@PathVariable String id, HttpHeaders headers) {
        Item item = itemService.findById(id);
        if (item == null) {
            return HttpResponse.notFound();
        }

        String eTag = itemService.calculateETag(item);
        if (headers.**ifNoneMatch**().contains(eTag)) {
            return HttpResponse.notModified().**eTag**(eTag);
        } else {
            return HttpResponse.ok(item).**eTag**(eTag);
        }
    }
}

Could see the highlighted ifNoneMatch and eTag are not getting recognised. Can you please help what am I missing in this simple code and why the eTag options are not getting accepted.

**Comment by graemerocher** [comment, marked as answer]
I assume you are coming from some other framework. These methods you are asking for are just short cuts to retrieving headers. Using headers.get(HttpHeaders.IF_MATCH) and HttpResponse.ok(item).header(HttpHeaders.ETAG, eTag) will achieve what you want.
